server/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -388,28 +388,35 @@ private InetSocketAddress bindToPort(final String name, final InetAddress hostAd
         PortsRange portsRange = new PortsRange(port);
         final AtomicReference<Exception> lastException = new AtomicReference<>();
         final AtomicReference<InetSocketAddress> boundSocket = new AtomicReference<>();
-        boolean success = portsRange.iterate(portNumber -> {
-            try {
-                TcpServerChannel channel = bind(name, new InetSocketAddress(hostAddress, portNumber));
-                synchronized (serverChannels) {
-                    List<TcpServerChannel> list = serverChannels.get(name);
-                    if (list == null) {
-                        list = new ArrayList<>();
-                        serverChannels.put(name, list);
+        closeLock.readLock().lock();
```

Comment:
@henningandersen @DaveCTurner shouldn't we acquire the `writeLock` here too to prevent this from running concurrently with the `doStop` (this method is only called from `doStart` anyway, so locking here shouldn't hurt anything since we're not concurrently)?
Otherwise, there isn't really anything that prevents a `put` to the `serverChannels` to run here if the shutdown happened after the above check for the `lifecycle` state passed is there? (which would lead us to still add to `serverChannels` after we're closed already)

Reply:
@original-brownbear the readLock here should prevent `doStop` from starting to close channels concurrently with this. Notice that lifecycle is moved to `STOPPED` before `doStop` is called and that the lifecycle state is a volatile, so this should be safe I think.

Reply:
@henningandersen Yea you're right <= Apparently I shouldn't have answered on a empty stomach.

Still, since this method is only called by `doStart` I'm wondering if the write lock isn't still the better choice since it would allow us to just remove the synchronization on `serverChannels` here and in `doStop` (those are the only two places touching this map) and simplify things a bit?
Just an optional idea though :)